File: dataset/stolen-vehicle-recovery/scripts/update_rto_images.py
import sqlite3
import urllib.request
import urllib.parse
import re
import json
import time

def search_ddg_image(query):
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'}
        url = f'https://duckduckgo.com/?q={urllib.parse.quote(query)}'
        req = urllib.request.Request(url, headers=headers)
        res = urllib.request.urlopen(req).read().decode('utf-8')
        vqd_match = re.search(r'vqd=([\d-]+)', res)
        if not vqd_match: return None
        vqd = vqd_match.group(1)
        
        api_url = f'https://duckduckgo.com/i.js?l=us-en&o=json&q={urllib.parse.quote(query)}&vqd={vqd}&f=,,,&p=1'
        req = urllib.request.Request(api_url, headers=headers)
        res = urllib.request.urlopen(req).read().decode('utf-8')
        data = json.loads(res)
        if 'results' in data and len(data['results']) > 0:
            return data['results'][0]['image']
    except Exception as e:
        print(f"Error searching DDG: {e}")
    return None

def main():
    print("Starting exact image matching using DuckDuckGo (Google Alternative)...")
    db_path = "/Users/prasanth/Documents/project v1/dataset/stolen-vehicle-recovery/prisma/dev.db"
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    cursor.execute("SELECT id, make, model, color, imageUrl FROM RTODatabase")
    records = cursor.fetchall()
    
    print(f"Found {len(records)} records. Processing...")
    
    updated = 0
    for record in records:
        record_id, make, model, color, current_url = record
        
        # We want to re-fetch to ensure the color is strictly matched
        query = f"{color} {make} {model} car high quality"
        print(f"Fetching image for: {color} {make} {model} ...")
        
        image_url = search_ddg_image(query)
        if image_url:
            cursor.execute("UPDATE RTODatabase SET imageUrl = ? WHERE id = ?", (image_url, record_id))
            print(f"  -> Success: {image_url}")
            updated += 1
            conn.commit()
        else:
            print("  -> No image found.")
        
        time.sleep(1) # Respect rate limits
        
    conn.close()
    print(f"Finished! Updated {updated} records.")

if __name__ == "__main__":
    main()

File: dataset/stolen-vehicle-recovery/scripts/update_rto_images.js
const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

async function fetchWikimediaImage(make, model, color) {
  try {
    // Search query prioritizing Make + Model
    // We add color if possible, but Wikipedia is unlikely to have color-specific car names as titles
    const query = encodeURIComponent(`${make} ${model}`);
    const url = `https://en.wikipedia.org/w/api.php?action=query&format=json&prop=pageimages&generator=search&gsrsearch=${query}&pithumbsize=500`;
    
    const response = await fetch(url);
    const data = await response.json();

    if (data.query && data.query.pages) {
      // Find the first valid page with an image
      const pages = Object.values(data.query.pages);
      // Sort by index to get the most relevant result
      pages.sort((a, b) => a.index - b.index);
      
      for (const page of pages) {
        if (page.thumbnail && page.thumbnail.source) {
          return page.thumbnail.source;
        }
      }
    }
  } catch (error) {
    console.error(`Failed to fetch image for ${make} ${model}:`, error.message);
  }
  return null;
}

async function main() {
  console.log('Starting RTO Database Image Seeding...');
  
  const records = await prisma.rTODatabase.findMany();
  console.log(`Found ${records.length} records. Processing...`);

  let updated = 0;
  for (const record of records) {
    if (!record.imageUrl) {
      console.log(`Fetching image for: ${record.make} ${record.model} (${record.color})`);
      const imageUrl = await fetchWikimediaImage(record.make, record.model, record.color);
      
      if (imageUrl) {
        await prisma.rTODatabase.update({
          where: { id: record.id },
          data: { imageUrl }
        });
        console.log(`  -> Success: ${imageUrl}`);
        updated++;
      } else {
        console.log(`  -> No image found.`);
      }
      
      // Sleep slightly to respect Wikipedia API limits
      await new Promise(resolve => setTimeout(resolve, 300));
    }
  }
  
  console.log(`Finished! Updated ${updated} records.`);
}

main()
  .catch(e => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
